src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -146,19 +156,66 @@ public HttpConnectionPoolManager(HttpConnectionSettings settings)
                         pool.OnNetworkChanged();
                     }
                 }
-                else
-                {
-                    // Our pools were GCed; user did not dispose the handler.
-                    NetworkChange.NetworkAddressChanged -= networkChangedDelegate;
-                }
             };
 
-            _networkChangedDelegate = networkChangedDelegate;
+            var cleanup = new NetworkChangeCleanup(networkChangedDelegate);
 
-            using (ExecutionContext.SuppressFlow())
+            if (Interlocked.CompareExchange(ref _networkChangeCleanup, cleanup, null) != null)
+            {
+                // We lost a race, someone else already started monitoring.
+                GC.SuppressFinalize(cleanup);
+                return;
+            }
+
+            bool restoreFlow = false;
+            try
             {
+                if (!ExecutionContext.IsFlowSuppressed())
+                {
+                    ExecutionContext.SuppressFlow();
+                    restoreFlow = true;
+                }
+
                 NetworkChange.NetworkAddressChanged += networkChangedDelegate;
             }
+            finally
+            {
+                if (restoreFlow) ExecutionContext.RestoreFlow();
+            }
+        }
+
+        private sealed class NetworkChangeCleanup : IDisposable
+        {
+            private NetworkAddressChangedEventHandler _handler;
+
+            public NetworkChangeCleanup(NetworkAddressChangedEventHandler handler)
+            {
+                _handler = handler;
+            }
+
+            // If user never disposes the HttpClient, use finalizer to remove from NetworkChange.NetworkAddressChanged.
+            ~NetworkChangeCleanup()
+            {
+                TryDispose();
+            }
+
+            public void Dispose()
+            {
+                if (TryDispose()) GC.SuppressFinalize(this);
```

Comment:
Can we simplify this to just always call GC.SuppressFinalize(this)?  Worst case is we call it multiple times if Dispose is called multiple times, right?

We also shouldn't need to null out _handler; worst case is we try to remove it multiple times, with subsequent times being a nop?

e.g. seems like it could be simplified to:
```C#
private sealed class NetworkChangeCleanup : IDisposable
{
    private readonly NetworkAddressChangedEventHandler _handler;

    public NetworkChangeCleanup(NetworkAddressChangedEventHandler handler) => _handler = handler;

    ~NetworkChangeCleanup() => NetworkChange.NetworkAddressChanged -= _handler;

    public void Dispose()
    {
        NetworkChange.NetworkAddressChanged -= _handler;
        GC.SuppressFinalize(this);
    }
}
```

Reply:
Sure.